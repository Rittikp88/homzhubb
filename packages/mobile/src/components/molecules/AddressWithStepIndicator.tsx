import React from 'react';
import { FlatList, StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Badge, Label, Text } from '@homzhub/common/src/components';
import { PropertyAddressCountry } from '@homzhub/mobile/src/components/molecules/PropertyAddressCountry';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { ILabelColor } from '@homzhub/common/src/domain/models/LeaseTransaction';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IProps {
  steps: string[];
  currentIndex: number;
  isStepDone: boolean[];
  primaryAddress: string;
  subAddress: string;
  propertyType: string;
  icon?: string;
  onEditPress?: () => void;
  selectedPan?: string;
  onPressSteps: (index: number) => void;
  badgeStyle?: StyleProp<ViewStyle>;
  stepContainerStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
}

export const AddressWithStepIndicator = (props: IProps): React.ReactElement => {
  const {
    steps,
    currentIndex,
    isStepDone,
    icon,
    selectedPan,
    badgeStyle,
    stepContainerStyle = {},
    containerStyle = {},
    primaryAddress,
    subAddress,
    propertyType,
    onPressSteps,
    onEditPress,
  } = props;

  const { t } = useTranslation(LocaleConstants.namespacesKey.property);

  const { rental, sell, informational, blue, green, darkTint7 } = theme.colors;

  const badgeData = (): ILabelColor => {
    switch (selectedPan) {
      case TypeOfPlan.RENT:
        return {
          label: t('forRental'),
          color: rental,
        };
      case TypeOfPlan.SELL:
        return {
          label: t('forSale'),
          color: sell,
        };
      case TypeOfPlan.MANAGE:
        return {
          label: t('manage'),
          color: informational,
        };
      default:
        return {
          label: t('forRental'),
          color: rental,
        };
    }
  };

  const renderIndicator = ({ item, index }: { item: string; index: number }): React.ReactElement => {
    const isCurrentStep = currentIndex === index;
    const stepDone = isStepDone[index];
    const iconColor = isCurrentStep ? blue : stepDone ? green : darkTint7;
    const labelStyle = [styles.stepLabel, isCurrentStep && styles.currentStepLabel, stepDone && styles.doneStepLabel];
    return (
      <View style={styles.indicatorView}>
        <Icon name={icons.roundFilled} size={30} color={iconColor} onPress={(): void => onPressSteps(index)} />
        <Label type="small" textType="semiBold" style={labelStyle}>
          {item}
        </Label>
      </View>
    );
  };

  const renderSeparator = (separatorData: { highlighted: boolean; leadingItem: string }): React.ReactElement => {
    let isDone = false;
    const indicatorSteps: string[] = steps;
    indicatorSteps.forEach((item, index) => {
      if (separatorData.leadingItem === item) {
        isDone = isStepDone[index];
      }
    });
    return <View style={[styles.separator, isDone && styles.doneSeparator]} />;
  };

  const badge = badgeData();

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={styles.topView}>
        <Text type="small" textType="regular">
          {propertyType}
        </Text>
        {icon && <Icon name={icon} size={23} color={blue} onPress={onEditPress} />}
        {selectedPan && <Badge title={badge.label.toUpperCase()} badgeColor={badge.color} badgeStyle={badgeStyle} />}
      </View>
      <PropertyAddressCountry
        primaryAddress={primaryAddress}
        subAddress={subAddress}
        containerStyle={styles.addressView}
      />
      <FlatList
        keyExtractor={(step): string => step}
        data={steps}
        renderItem={renderIndicator}
        horizontal
        showsHorizontalScrollIndicator={false}
        ItemSeparatorComponent={renderSeparator}
        style={[styles.listStyle, stepContainerStyle]}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 16,
    borderRadius: 4,
    backgroundColor: theme.colors.white,
  },
  addressView: {
    marginVertical: 6,
  },
  indicatorView: {
    alignItems: 'center',
  },
  topView: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  separator: {
    backgroundColor: theme.colors.darkTint7,
    width: 50,
    height: 2,
    marginTop: 10,
    opacity: 0.3,
  },
  doneSeparator: {
    backgroundColor: theme.colors.green,
  },
  listStyle: {
    marginTop: 10,
    alignSelf: 'center',
  },
  stepLabel: {
    color: theme.colors.darkTint7,
  },
  currentStepLabel: {
    color: theme.colors.blue,
  },
  doneStepLabel: {
    color: theme.colors.green,
  },
});
