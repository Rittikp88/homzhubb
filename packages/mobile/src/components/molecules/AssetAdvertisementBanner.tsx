import React from 'react';
import { Image, StyleSheet, View } from 'react-native';
import { DashboardRepository } from '@homzhub/common/src/domain/repositories/DashboardRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
import { PaginationComponent } from '@homzhub/mobile/src/components/atoms/PaginationComponent';
import { AssetAdvertisement, AssetAdvertisementResults } from '@homzhub/common/src/domain/models/AssetAdvertisement';

interface IAssetAdvertisementBannerState {
  banners: AssetAdvertisement;
  activeSlide: number;
}

export class AssetAdvertisementBanner extends React.PureComponent<{}, IAssetAdvertisementBannerState> {
  public state = {
    banners: {} as AssetAdvertisement,
    activeSlide: 0,
  };

  public componentDidMount = async (): Promise<void> => {
    await this.fetchAdvertisementBanners();
  };

  public render(): React.ReactNode {
    const { banners, activeSlide } = this.state;
    if (!banners) {
      return null;
    }
    return (
      <View style={styles.container}>
        <SnapCarousel
          carouselData={banners?.results ?? []}
          carouselItem={this.renderCarouselItem}
          itemWidth={theme.viewport.width}
          activeIndex={activeSlide}
          onSnapToItem={this.onSnapToItem}
          testID="bannerSnap"
        />
        <PaginationComponent
          dotsLength={banners?.results?.length ?? 0}
          activeSlide={activeSlide}
          containerStyle={styles.overlay}
          activeDotStyle={styles.activeDotStyle}
          inactiveDotStyle={styles.inactiveDotStyle}
        />
      </View>
    );
  }

  private renderCarouselItem = (item: AssetAdvertisementResults): React.ReactElement => {
    return (
      <Image
        source={{
          uri: item.attachment.link,
        }}
        style={styles.carouselImage}
      />
    );
  };

  public onSnapToItem = (slideNumber: number): void => {
    this.setState({ activeSlide: slideNumber });
  };

  public fetchAdvertisementBanners = async (): Promise<void> => {
    const response: AssetAdvertisement = await DashboardRepository.getAdvertisements();
    this.setState({ banners: response });
  };
}

const styles = StyleSheet.create({
  container: {
    borderRadius: 4,
    height: 180,
    overflow: 'hidden',
  },
  overlay: {
    position: 'absolute',
    flex: 0,
    alignSelf: 'center',
    bottom: 0,
    paddingVertical: 5,
  },
  carouselImage: {
    height: '100%',
    width: '100%',
  },
  activeDotStyle: {
    borderWidth: 1,
  },
  inactiveDotStyle: {
    backgroundColor: theme.colors.darkTint6,
    borderColor: theme.colors.transparent,
  },
});
