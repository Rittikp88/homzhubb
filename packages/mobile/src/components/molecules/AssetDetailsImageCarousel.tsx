import React from 'react';
import { View, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import { ImageVideoPagination } from '@homzhub/common/src/components';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';

interface IProps {
  images: any;
  enterFullScreen: () => void;
  updateSlide: (index: number) => void;
  activeSlide: number;
}

export class AssetDetailsImageCarousel extends React.PureComponent<IProps> {
  public render(): React.ReactElement {
    const { activeSlide, images } = this.props;
    return (
      <View style={styles.carouselContainer}>
        <SnapCarousel
          carouselData={images}
          carouselItem={this.renderCarouselItem}
          activeIndex={activeSlide}
          sliderWidth={theme.viewport.width}
          itemWidth={theme.viewport.width}
          onSnapToItem={this.onSnapToItem}
        />
        <View style={styles.overlay}>
          <ImageVideoPagination
            currentSlide={activeSlide}
            totalSlides={images.length}
            type={images[activeSlide].attachment_type}
          />
        </View>
      </View>
    );
  }

  private renderCarouselItem = (item: any): React.ReactElement => {
    const { enterFullScreen } = this.props;
    return (
      <TouchableOpacity onPress={enterFullScreen}>
        {item.attachment_type === 'IMAGE' && (
          <Image
            source={{
              uri: item.link,
            }}
            style={styles.carouselImage}
          />
        )}
        {/* TODO: Remove once the api is integrated */}
        {item.attachment_type === 'VIDEO' && <Image source={item.thumbnail_url} style={styles.carouselImage} />}
      </TouchableOpacity>
    );
  };

  public onSnapToItem = (slideNumber: number): void => {
    const { updateSlide } = this.props;
    updateSlide(slideNumber);
  };
}

const styles = StyleSheet.create({
  carouselContainer: {
    borderRadius: 4,
    overflow: 'hidden',
    backgroundColor: theme.colors.darkTint1,
  },
  overlay: {
    position: 'absolute',
    alignSelf: 'center',
    bottom: 15,
    backgroundColor: theme.colors.carouselCardOpacity,
  },
  carouselImage: {
    height: '100%',
    width: '100%',
  },
});
