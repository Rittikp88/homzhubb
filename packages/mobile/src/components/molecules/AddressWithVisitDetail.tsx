import React from 'react';
import { StyleProp, StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { PropertyAddress } from '@homzhub/common/src/components/molecules/PropertyAddress';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { VisitSlot } from '@homzhub/common/src/mocks/BookVisit';

interface IProps {
  primaryAddress: string;
  subAddress: string;
  startDate: string;
  endDate: string;
  comments?: string;
  isMissedVisit?: boolean;
  isCompletedVisit?: boolean;
  isRescheduleAll?: boolean;
  isFromProperty?: boolean;
  onPressSchedule?: () => void;
  containerStyle?: StyleProp<ViewStyle>;
}

export const AddressWithVisitDetail = (props: IProps): React.ReactElement => {
  const {
    subAddress,
    primaryAddress,
    startDate,
    comments,
    isMissedVisit,
    isCompletedVisit,
    containerStyle = {},
    onPressSchedule,
    isRescheduleAll = false,
    isFromProperty = false,
  } = props;
  const { t } = useTranslation(LocaleConstants.namespacesKey.property);
  const dateTime = DateUtils.convertTimeFormat(startDate, 'YYYY-MM-DD HH');
  const time = VisitSlot.find((item) => item.from === Number(dateTime[1]));
  const textStyle = [styles.textColor, isMissedVisit && styles.missedColor];
  return (
    <View style={containerStyle}>
      {isFromProperty ? (
        <>
          {!!comments && (
            <Text type="small" textType="semiBold" style={styles.addressContainer}>
              {comments}
            </Text>
          )}
        </>
      ) : (
        <PropertyAddress
          primaryTextType="small"
          subAddress={subAddress}
          primaryAddress={primaryAddress}
          subAddressStyle={styles.subAddress}
          containerStyle={styles.addressContainer}
        />
      )}
      <Label type="regular" style={[styles.textColor, !comments && styles.detail]}>
        {t('visitDetails')}
      </Label>
      <View style={styles.detailContainer}>
        <View style={styles.content}>
          <Text type="small" textType="semiBold" style={textStyle}>
            {DateUtils.getDisplayDate(dateTime[0], 'DD MMM')}
          </Text>
          <Icon
            name={icons.roundFilled}
            color={isMissedVisit ? theme.colors.error : theme.colors.darkTint3}
            size={8}
            style={styles.iconStyle}
          />
          <Text type="small" textType="semiBold" style={textStyle}>
            {time?.formatted}
          </Text>
        </View>
        <TouchableOpacity style={styles.content} onPress={onPressSchedule}>
          <Icon name={icons.schedule} color={theme.colors.blue} size={20} />
          <Text type="small" style={styles.scheduleText}>
            {isCompletedVisit ? t('newVisit') : isRescheduleAll ? t('rescheduleAll') : t('reschedule')}
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  subAddress: {
    marginLeft: 0,
  },
  addressContainer: {
    marginVertical: 12,
  },
  textColor: {
    color: theme.colors.darkTint3,
  },
  detail: {
    marginTop: 12,
  },
  missedColor: {
    color: theme.colors.error,
  },
  detailContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconStyle: {
    marginTop: 6,
    marginHorizontal: 6,
  },
  scheduleText: {
    color: theme.colors.blue,
    marginStart: 6,
  },
});
