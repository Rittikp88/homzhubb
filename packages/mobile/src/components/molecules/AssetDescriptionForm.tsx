import React from 'react';
import { View, StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { FormikProps, FormikValues } from 'formik';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormCalendar, FormDropdown, FormTextInput, Text } from '@homzhub/common/src/components';
import { FormCounter } from '@homzhub/common/src/components/molecules/FormCounter';
import { AssetDescriptionDropdownValues } from '@homzhub/common/src/domain/models/AssetDescriptionForm';

interface IProps {
  formProps: FormikProps<FormikValues>;
  dropDownOptions: AssetDescriptionDropdownValues;
}

const AssetDescriptionForm = ({ formProps, dropDownOptions }: IProps): React.ReactElement => {
  const [t] = useTranslation(LocaleConstants.namespacesKey.property);
  // Todo (Sriram: Look into this prefill issue of carpetArea)

  return (
    <>
      <Text type="small" style={styles.headingStyle}>
        {t('assetDescription:description')}
      </Text>
      <View style={styles.formContainer}>
        <View style={styles.contentView}>
          <View style={styles.subContentView}>
            <FormTextInput
              style={styles.inputFieldStyle}
              name="carpetArea"
              label={t('carpetArea')}
              maxLength={10}
              numberOfLines={1}
              inputType="number"
              placeholder={t('common:enter')}
              formProps={formProps}
            />
          </View>
          <View style={styles.flexOne}>
            <FormDropdown
              name="areaUnit"
              label={t('areaUnit')}
              options={dropDownOptions.areaUnitDropdownValues}
              placeholder={t('selectAreaUnit')}
              maxLabelLength={8}
              formProps={formProps}
            />
          </View>
        </View>
        <FormDropdown
          label={t('facingText')}
          name="facing"
          options={dropDownOptions.facing}
          placeholder={t('propertySearch:selectFacing')}
          maxLabelLength={36}
          formProps={formProps}
        />
        <FormDropdown
          label={t('flooringType')}
          name="flooringType"
          options={dropDownOptions.typeOfFlooring}
          placeholder={t('selectFlooringType')}
          maxLabelLength={36}
          formProps={formProps}
        />
        <FormCalendar
          formProps={formProps}
          label={t('yearOfConstruction')}
          name="yearOfConstruction"
          allowPastDates
          isYearView
          placeHolder={t('assetFinancial:addDatePlaceholder')}
          placeHolderStyle={styles.placeHolderStyle}
          dateStyle={styles.dateStyle}
          textType="label"
        />
        <FormCounter
          containerStyles={styles.marginTop}
          name="totalFloors"
          maxCount={1000}
          label={t('numberOfFloorsText')}
          formProps={formProps}
        />
        <FormCounter
          containerStyles={styles.marginTop}
          name="onFloorNumber"
          label={t('onFloorText')}
          formProps={formProps}
          maxCount={formProps.values.totalFloors}
        />
      </View>
    </>
  );
};

const memoizedComponent = React.memo(AssetDescriptionForm);
export { memoizedComponent as AssetDescriptionForm };

const styles = StyleSheet.create({
  headingStyle: {
    marginTop: 16,
    paddingVertical: 16,
    paddingLeft: 16,
    backgroundColor: theme.colors.moreSeparator,
  },
  formContainer: {
    backgroundColor: theme.colors.white,
    paddingHorizontal: 24,
    paddingBottom: 24,
  },
  contentView: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  subContentView: {
    flex: 1,
    marginRight: 16,
  },
  flexOne: {
    flex: 1,
  },
  marginTop: {
    marginTop: 24,
  },
  inputFieldStyle: {
    height: 40,
  },
  dateStyle: {
    marginLeft: 0,
  },
  placeHolderStyle: {
    color: theme.colors.darkTint8,
  },
});
