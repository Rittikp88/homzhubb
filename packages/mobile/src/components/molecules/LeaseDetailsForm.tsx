import React from 'react';
import { StyleSheet } from 'react-native';
import { Formik, FormikActions, FormikProps, FormikValues } from 'formik';
import { withTranslation, WithTranslation } from 'react-i18next';
import moment from 'moment';
import * as yup from 'yup';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormTextInput, Text, FormButton, Slider } from '@homzhub/common/src/components';
import { ButtonGroup } from '@homzhub/mobile/src/components/molecules/ButtonGroup';
import { FormCalendar } from '@homzhub/common/src/components/molecules/FormCalendar';
import { MaintenanceDetails } from '@homzhub/mobile/src/components/molecules/MaintenanceDetails';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import {
  FurnishingType,
  ICreateLeaseTermDetails,
  ILeaseTermDetails,
  PaidByTypes,
  ScheduleTypes,
} from '@homzhub/common/src/domain/models/LeaseTerms';

interface IState {
  showMore: boolean;
  monthlyRent: string;
  securityDeposit: string;
  annualIncrement: string;
  maintenanceAmount: string;
  maintenanceSchedule: ScheduleTypes;
  minimumLeasePeriod: number;
  availableFrom: string;
  furnishingStatus: FurnishingType;
  maintenanceBy: PaidByTypes;
  utilityBy: PaidByTypes;
}

interface IProps extends WithTranslation {
  initialValues: ILeaseTermDetails | null;
  currency: string;
  onSubmit: (data: ICreateLeaseTermDetails) => void;
}

export enum LeaseFormKeys {
  monthlyRent = 'monthlyRent',
  securityDeposit = 'securityDeposit',
  showMore = 'showMore',
  annualIncrement = 'annualIncrement',
  maintenanceAmount = 'maintenanceAmount',
  maintenanceSchedule = 'maintenanceSchedule',
  maintenanceBy = 'maintenanceBy',
  availableFrom = 'availableFrom',
  furnishingStatus = 'furnishingStatus',
  utilityBy = 'utilityBy',
  minimumLeasePeriod = 'minimumLeasePeriod',
}
const MINIMUM_LEASE_PERIOD = 1;
const DEFAULT_LEASE_PERIOD = 11;
const MAXIMUM_LEASE_PERIOD = 24;

class LeaseDetailsForm extends React.PureComponent<IProps, IState> {
  /*eslint-disable */
  private PAID_BY_OPTIONS = [
    { title: this.props.t('owner'), value: PaidByTypes.OWNER },
    { title: this.props.t('tenant'), value: PaidByTypes.TENANT },
  ];
  private FURNISHING_STATUS = [
    { title: this.props.t('fullyFurnished'), value: FurnishingType.FULL },
    { title: this.props.t('semiFurnished'), value: FurnishingType.SEMI },
    { title: this.props.t('none'), value: FurnishingType.NONE },
  ];
  /* eslint-enable */

  public state = {
    showMore: false,
    monthlyRent: '',
    securityDeposit: '',
    annualIncrement: '',
    maintenanceAmount: '',
    maintenanceSchedule: ScheduleTypes.ANNUALLY,
    minimumLeasePeriod: DEFAULT_LEASE_PERIOD,
    availableFrom: moment().format('YYYY-MM-DD'),
    furnishingStatus: FurnishingType.NONE,
    maintenanceBy: PaidByTypes.OWNER,
    utilityBy: PaidByTypes.TENANT,
  };

  public componentDidUpdate = (prevProps: Readonly<IProps>, prevState: Readonly<IState>): void => {
    const { initialValues } = this.props;
    if (prevProps.initialValues !== initialValues && initialValues) {
      const {
        maintenance_schedule,
        monthly_rent_price,
        security_deposit_price,
        annual_increment_percentage,
        maintenance_amount,
        minimum_lease_period,
        available_from_date,
        furnishing_status,
        maintenance_paid_by,
        utility_paid_by,
      } = initialValues;
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        showMore: !!annual_increment_percentage,
        monthlyRent: monthly_rent_price.toString(),
        securityDeposit: security_deposit_price.toString(),
        annualIncrement: annual_increment_percentage?.toString() ?? '',
        maintenanceAmount: maintenance_amount?.toString() ?? '',
        maintenanceSchedule: maintenance_schedule ?? ScheduleTypes.ANNUALLY,
        minimumLeasePeriod: minimum_lease_period,
        availableFrom: available_from_date,
        furnishingStatus: furnishing_status,
        maintenanceBy: maintenance_paid_by,
        utilityBy: utility_paid_by,
      });
    }
  };

  public render(): React.ReactNode {
    const { t, currency } = this.props;
    return (
      <>
        <Formik
          enableReinitialize
          onSubmit={this.onSubmit}
          initialValues={{ ...this.state }}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<IState>): React.ReactElement => {
            const { setFieldValue, setFieldTouched, values } = formProps;
            const onShowMorePress = (): void => {
              setFieldValue(LeaseFormKeys.showMore, !values.showMore);
              if (!values.showMore) {
                setFieldTouched(LeaseFormKeys.annualIncrement, false);
              }
            };
            const onMaintenanceChanged = (value: PaidByTypes): void => {
              setFieldValue(LeaseFormKeys.maintenanceBy, value);
              if (value === PaidByTypes.TENANT) {
                setFieldTouched(LeaseFormKeys.maintenanceAmount, false);
              }
            };

            return (
              <>
                <FormTextInput
                  inputType="number"
                  name={LeaseFormKeys.monthlyRent}
                  label={t('monthlyRent')}
                  placeholder={t('monthlyRentPlaceholder')}
                  maxLength={12}
                  formProps={formProps}
                  inputGroupSuffixText={currency}
                />
                <FormTextInput
                  inputType="number"
                  name={LeaseFormKeys.securityDeposit}
                  label={t('securityDeposit')}
                  placeholder={t('securityDepositPlaceholder')}
                  maxLength={12}
                  formProps={formProps}
                  inputGroupSuffixText={currency}
                />
                <Text type="small" textType="semiBold" style={styles.showMore} onPress={onShowMorePress}>
                  {values.showMore ? t('showLess') : t('showMore')}
                </Text>
                {values.showMore && (
                  <FormTextInput
                    inputType="decimal"
                    name={LeaseFormKeys.annualIncrement}
                    label={t('annualIncrement')}
                    placeholder={t('annualIncrementPlaceholder')}
                    maxLength={4}
                    formProps={formProps}
                    inputGroupSuffixText={t('annualIncrementSuffix')}
                  />
                )}
                {this.renderNonFormikInputs(formProps)}
                <Text type="small" textType="semiBold" style={styles.headerTitle}>
                  {t('maintenanceBy')}
                </Text>
                <ButtonGroup<PaidByTypes>
                  data={this.PAID_BY_OPTIONS}
                  onItemSelect={onMaintenanceChanged}
                  selectedItem={values.maintenanceBy}
                />
                {values.maintenanceBy === PaidByTypes.TENANT && (
                  <MaintenanceDetails
                    formProps={formProps}
                    currency={currency}
                    maintenanceAmountKey={LeaseFormKeys.maintenanceAmount}
                    maintenanceScheduleKey={LeaseFormKeys.maintenanceSchedule}
                  />
                )}
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  formProps={formProps}
                  type="primary"
                  title={t('common:continue')}
                  containerStyle={styles.continue}
                />
              </>
            );
          }}
        </Formik>
      </>
    );
  }

  private renderNonFormikInputs = (formProps: FormikProps<IState>): React.ReactNode => {
    const { t } = this.props;
    const { minimumLeasePeriod } = this.state;
    const { values, setFieldValue } = formProps;

    const onSliderChange = (value: number): void => {
      setFieldValue(LeaseFormKeys.minimumLeasePeriod, Math.round(value));
    };

    const onFurnishingStatusChange = (value: string): void => {
      setFieldValue(LeaseFormKeys.furnishingStatus, value);
    };

    const onUtilityChanged = (value: string): void => {
      setFieldValue(LeaseFormKeys.utilityBy, value);
    };

    return (
      <>
        <Text type="small" textType="semiBold" style={styles.headerTitle}>
          {t('duration')}
        </Text>
        <FormCalendar name="availableFrom" formProps={formProps} containerStyle={styles.dateContainer} />
        <>
          <Text type="small" textType="semiBold" style={styles.headerTitle}>
            {t('minimumLeasePeriod')}
          </Text>
          <Slider
            onSliderChange={onSliderChange}
            minSliderRange={MINIMUM_LEASE_PERIOD}
            maxSliderRange={MAXIMUM_LEASE_PERIOD}
            minSliderValue={minimumLeasePeriod}
            isLabelRequired
            labelText="Months"
          />
        </>
        <Text type="small" textType="semiBold" style={styles.headerTitle}>
          {t('furnishing')}
        </Text>
        <ButtonGroup<FurnishingType>
          data={this.FURNISHING_STATUS}
          onItemSelect={onFurnishingStatusChange}
          selectedItem={values[LeaseFormKeys.furnishingStatus]}
        />
        <Text type="small" textType="semiBold" style={styles.headerTitle}>
          {t('utilityBy')}
        </Text>
        <ButtonGroup<PaidByTypes>
          data={this.PAID_BY_OPTIONS}
          onItemSelect={onUtilityChanged}
          selectedItem={values[LeaseFormKeys.utilityBy]}
        />
      </>
    );
  };

  private onSubmit = (values: IState, formActions: FormikActions<FormikValues>): void => {
    formActions.setSubmitting(true);
    const { onSubmit, currency } = this.props;

    let maintenance_amount: number | null = parseInt(values[LeaseFormKeys.maintenanceAmount], 10);
    let maintenance_schedule: ScheduleTypes | null = values[LeaseFormKeys.maintenanceSchedule];
    let annual_increment_percentage: number | null = parseFloat(values[LeaseFormKeys.annualIncrement]);
    if (values[LeaseFormKeys.maintenanceBy] === PaidByTypes.OWNER) {
      maintenance_amount = null;
      maintenance_schedule = null;
    }
    if (!values[LeaseFormKeys.showMore]) {
      annual_increment_percentage = null;
    }

    const leaseTerms: ICreateLeaseTermDetails = {
      currency_code: currency,
      monthly_rent_price: parseInt(values[LeaseFormKeys.monthlyRent], 10),
      security_deposit_price: parseInt(values[LeaseFormKeys.securityDeposit], 10),
      annual_increment_percentage,
      minimum_lease_period: values[LeaseFormKeys.minimumLeasePeriod],
      available_from_date: values[LeaseFormKeys.availableFrom],
      maintenance_paid_by: values[LeaseFormKeys.maintenanceBy],
      utility_paid_by: values[LeaseFormKeys.utilityBy],
      furnishing_status: values[LeaseFormKeys.furnishingStatus],
      maintenance_amount,
      maintenance_schedule,
    };
    onSubmit(leaseTerms);
    formActions.setSubmitting(false);
  };

  private formSchema = (): yup.ObjectSchema => {
    const { t } = this.props;
    return yup.object().shape({
      [LeaseFormKeys.maintenanceBy]: yup.string(),
      [LeaseFormKeys.showMore]: yup.boolean(),
      [LeaseFormKeys.monthlyRent]: yup
        .string()
        .matches(FormUtils.digitRegex, t('common:onlyNumeric'))
        .required(t('monthlyRentRequired')),
      [LeaseFormKeys.securityDeposit]: yup
        .string()
        .matches(FormUtils.digitRegex, t('common:onlyNumeric'))
        .required(t('securityDepositRequired')),
      [LeaseFormKeys.annualIncrement]: yup.string().when('showMore', {
        is: true,
        then: yup
          .string()
          .matches(FormUtils.percentageRegex, t('common:onlyNumeric'))
          .required(t('annualIncrementRequired')),
      }),
      [LeaseFormKeys.maintenanceAmount]: yup.string().when('maintenanceBy', {
        is: PaidByTypes.TENANT,
        then: yup
          .string()
          .matches(FormUtils.digitRegex, t('common:onlyNumeric'))
          .required(t('maintenanceAmountRequired')),
      }),
      [LeaseFormKeys.maintenanceSchedule]: yup.string<ScheduleTypes>().required(t('maintenanceScheduleRequired')),
    });
  };
}

const HOC = withTranslation(LocaleConstants.namespacesKey.property)(LeaseDetailsForm);
export { HOC as LeaseDetailsForm };

const styles = StyleSheet.create({
  continue: {
    flex: 0,
    marginVertical: 28,
  },
  showMore: {
    marginTop: 8,
    alignSelf: 'flex-end',
    color: theme.colors.active,
  },
  headerTitle: {
    marginTop: 28,
    marginBottom: 16,
    color: theme.colors.darkTint3,
  },
  dateContainer: {
    marginTop: 0,
  },
});
