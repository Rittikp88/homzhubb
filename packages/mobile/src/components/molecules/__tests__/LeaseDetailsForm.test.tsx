import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { LeaseDetailsForm } from '@homzhub/mobile/src/components/molecules/LeaseDetailsForm';
import { leaseTermDetail } from '@homzhub/common/src/mocks/PropertyDetails';

let props: any;
const mockFunction = jest.fn();

describe('LeaseDetailsForm', () => {
  const createTestProps = (testProps: any): object => ({
    initialValues: leaseTermDetail,
    currency: '',
    onSubmit: mockFunction,
    ...testProps,
  });
  props = createTestProps({});

  const wrapper = shallow(<LeaseDetailsForm {...props} t={(key: string): string => key} />);

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
