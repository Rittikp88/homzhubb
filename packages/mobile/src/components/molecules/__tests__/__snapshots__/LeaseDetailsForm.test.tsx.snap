// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseDetailsForm should match snapshot 1`] = `
<Fragment>
  <Formik
    enableReinitialize={true}
    initialValues={
      Object {
        "annualIncrement": "",
        "availableFrom": "2020-09-08",
        "furnishingStatus": "NONE",
        "maintenanceAmount": "",
        "maintenanceBy": "OWNER",
        "maintenanceSchedule": "ANNUALLY",
        "minimumLeasePeriod": 11,
        "monthlyRent": "",
        "securityDeposit": "",
        "showMore": false,
        "utilityBy": "TENANT",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Component />
  </Formik>
</Fragment>
`;
